Enable isolation groups for Java 9+

pk-work opened the pull request:
The current implementation of isolation groups [1] requires the
instantiation of a new IsolatingClassLoader, which contains all JARs
from the class path of the SystemClassLoader. Before Java 9, this was
solved by casting the SystemClassLoader to an URLClassLoader to extract
JARs on the class path of the SystemClassLoader.

Due to the fact, that the SystemClassLoader is not longer an
URLClassLoader in Java 9, this approach will not longer work.

This patch implements a way to extract the class path of the
SystemClassLoader which does not require the SystemClassLoader to be an
instance of URLClassLoader.

WARNING: This approach will not work for modules, which were introduces
in Java 9, because modules are living in the "runtime image" [2]. Due to
this, it is not possible to pass the module classes into the current
implementation of IsolatingClassLoader.


[1] https://github.com/eclipse-vertx/vert.x/blob/4bbced9e8ac173ab73c4b4d1ef0bee99c6322a89/src/main/asciidoc/index.adoc#verticle-isolation-groups
[2] https://openjdk.java.net/jeps/220


Signed-off-by: Pascal Krause <pascal.krause@sap.com>

pk-work commented:
This is a patch for issue #2278

pk-work commented:
When I run the tests on my machine they failed, even without my change. I will try to get the test running on my local machine. Could someone else test this change with Java 9+, too? Or could we configure Travis, that Travis runs this test with Java 9+?

vietj commented:
@pk-work what does fail for you ?

vietj commented:
in addition this PR should be verified on Java 11 by the person merging as the CI does not run the job for java > 8

pk-work commented:
After I deleted and cloned the repository again I managed to build and run this change with Java 8. Maybe some remnant of my research.

After that I changed in the pom.xml the following lines:
```
     <profile>
-      <id>java-8</id>
+      <id>java-11</id>
     <properties>
```

```  
   <activation>
-        <jdk>1.8</jdk>
+        <jdk>11</jdk>
    </activation>
```
When I ran the tests again, I got the following error:

```
Failed tests: 
  ClusteredEventBusStartFailureTest.testCallbackInvokedOnFailure:47->AsyncTestBase.awaitLatch:596->AsyncTestBase.assertTrue:376 null

Tests run: 3723, Failures: 1, Errors: 0, Skipped: 12
```

vietj commented:
that test is independant from your contribution, some tests are badly written and sometimes fail (and we need to improve them)

pk-work commented:
I ran the tests now several times with different Java versions, this is the outcome:

Java  9 - 2 runs - 2 succeeded
Java 10 + 3 runs - 2 failed, 1 succeeded

```
Tests in error: 
  ClusteredEventBusWithSSLTest>AsyncTestBase.after:81->VertxTestBase.tearDown:85->AsyncTestBase.close:653 » UnsupportedOperation
  Http1xTest.testCloseTheConnectionAfterResetBeforeKeepAliveResponseReceived:3479->testCloseTheConnectionAfterResetBeforeResponseReceived:3569->AsyncTestBase.await:119->AsyncTestBase.await:131 » IllegalState
  Http1xTest>AsyncTestBase.after:81->HttpTestBase.tearDown:65->VertxTestBase.tearDown:85->AsyncTestBase.close:653 » UnsupportedOperation
  ClusteredSharedCounterTest>AsyncTestBase.after:81->VertxTestBase.tearDown:85->AsyncTestBase.close:653 » UnsupportedOperation

Tests run: 3723, Failures: 0, Errors: 4, Skipped: 12
```

java 11 - 2 runs - 1 failed, 1 succeeded

```
Tests in error: 
  Http1xTest.testCloseTheConnectionAfterResetBeforeKeepAliveResponseReceived:3479->testCloseTheConnectionAfterResetBeforeResponseReceived:3569->AsyncTestBase.await:119->AsyncTestBase.await:131 » IllegalState

Tests run: 3723, Failures: 0, Errors: 1, Skipped: 12
```

My observation was: If you look at the console all the time during the tests, they run through. Maybe they just need a little attention to work :)

The only downside of this patch is, that folders on class path (e.g. <pathToMyEclipseWorkSpace>/Test/bin/), will be ignored. Maybe this should be documented that with Java 9+ only classes in JAR files will be recognized.

vietj commented:
the UnsupportedOperation come from Netty itself.

https://github.com/netty/netty/issues/8475

> On 14 Dec 2018, at 14:00, Pascal <notifications@github.com> wrote:
> 
> UnsupportedOperation

pk-work commented:
Should I update also the documentation? Or is this limitation okay for you?

pk-work commented:
I just found a bug in the change .. Which shows that a test is missing.

I will provide a test case for this.

pk-work commented:
retest this please

pk-work commented:
Close / Reopen to retest

cescoffier commented on src/main/java/io/vertx/core/impl/DeploymentManager.java:
```diff
@@ -374,19 +376,68 @@ public void unregisterVerticleFactory(VerticleFactory factory) {
+  private static List<URL> extractCPFromProperty() {
+    String classpath = System.getProperty("java.class.path");
+    List<URL> classpathURLs = new ArrayList<>();
+    for (String path : classpath.split(":")) {
```
Use path separator, this can break depending on OS.

pk-work replied:
Done

cescoffier commented on src/main/java/io/vertx/core/impl/DeploymentManager.java:
```diff
@@ -374,19 +376,68 @@ public void unregisterVerticleFactory(VerticleFactory factory) {
+  }
+
+  private static List<URL> extractCPFromProperty() {
+    String classpath = System.getProperty("java.class.path");
```
In some rare circumstances, this can be `null`

pk-work replied:
Done

cescoffier commented on src/test/java/io/vertx/core/impl/DeploymentManagerTest.java:
```diff
@@ -0,0 +1,20 @@
+  public void extractClasspathTest() throws MalformedURLException, URISyntaxException {
+    for(URL cpURL : DeploymentManager.extractClasspath(getClass().getClassLoader())) {
+      Assert.assertTrue(cpURL + " is not a valid classpath entry", new File(cpURL.toURI()).exists());
+    }
```
Add a test with an exploded classpath, it may behave differently.

pk-work replied:
Done, good catch, this resulted really in a problem :)

cescoffier commented on src/main/java/io/vertx/core/impl/DeploymentManager.java:
```diff
@@ -374,19 +376,68 @@ public void unregisterVerticleFactory(VerticleFactory factory) {
+      throw new IllegalStateException(e);
+    }
+    for (URL url : Collections.list(urls)) {
+      String path = url.getFile().replace(prefix, "").replace(suffix, "").trim();
```
This is specific, typically in OSGi there are not `file:`.

pk-work replied:
But if there is not "file:", it will just not replace it. Or do you mean, there could be something else and  I should delete everything from beginning to the first colon?

pk-work replied:
I now fixed it, but not sure if this is a proper solution.

pk-work commented:
@cescoffier  Thank you so much for your valuable feedback. I think I provided a fix for your requests. Except the last comment was not 100% clear to me. Maybe you can have a look at my question.

pk-work commented on src/main/java/io/vertx/core/impl/DeploymentManager.java:
```diff
@@ -374,19 +377,80 @@ public void unregisterVerticleFactory(VerticleFactory factory) {
 
+  private static URL mapToURL(String path) {
+    try {
+      if (path.startsWith("/")) {
```
This will not work for windows.
Test for this regexp ^(/|[a-zA-Z]\\:\\\\).* will work for absolute paths but not for relative paths.

But how to check if the string starts with a protocol or an relative path, because at least in linux a colon is a valid char in in the path ....

Any ideas?

pk-work replied:
Okay, I found a way to solve this problem:

The only place where this method is called with strings containing an URL is in extractCPByManifest(). I will not longer call mapToURL from this place, then I can expect always a String which represents an absolute file path, which can be handled by new File()...

vietj commented:
@cescoffier is this good to merge ?

cescoffier commented:
should be good right now.

pk-work commented:
Hi, we now really uses this patch now in our internal fork and I found a stupid bug. I will upload a new patch soon. I simply forgot to remove the check if the SystemClassLoader is a URLClassLoader.

I'm so happy that this is not already merged. Sorry for that stupid mistake.

amaceing commented:
Do you know when/if this will get merged and published?

vietj commented:
the PR should be done on the 3.7 branch, not master

vietj commented:
@pk-work can you redo the PR on 3.7 branch ?

pk-work commented:
Done. Don't know why there are still conflicting files, I rebased on 3.7. Should be outdated.

vietj commented:
can you make a new clean PR ?

pk-work commented:
Done #2872
